Python/Product/PythonTools/PythonTools/Editor/PythonTextBufferInfo.cs

```diff
@@ -472,31 +473,20 @@ sealed class PythonTextBufferInfo {
             int startCol = span.Start - span.Start.GetContainingLine().Start;
             int endCol = span.End - span.End.GetContainingLine().Start;
 
-
-            // Cache may change from background thread during enumeration
-            // so make sure we track that snapshot version is still the same.
-            var spanSnapshot = span.Snapshot;
-
-            Tokenizer tokenizer = null;
-            for (int line = firstLine; line <= lastLine; ++line) {
-                var currentBufferSpanshot = span.Snapshot.TextBuffer.CurrentSnapshot;
-                if (spanSnapshot.Version != currentBufferSpanshot.Version) {
-                    yield break;
-                }
-
-                var lineTokenization = _tokenCache.GetLineTokenization(span.Snapshot.GetLineFromLineNumber(line), () => tokenizer ?? (tokenizer = MakeTokenizer()));
-
-                foreach (var token in lineTokenization.Tokens.MaybeEnumerate()) {
-                    if (line < firstLine || line > lastLine) {
-                        continue;
-                    }
-                    if (line == firstLine && token.Column + token.Length < startCol) {
-                        continue;
+            // We need current state of the cache since it can change from a background thread
+            using (var cacheSnapshot = _tokenCache.GetSnapshot()) {
+                for (int line = firstLine; line <= lastLine; ++line) {
+                    var lineTokenization = cacheSnapshot.GetLineTokenization(span.Snapshot.GetLineFromLineNumber(line), GetTokenizerLazy());
```

Comment:
Calling `GetTokenizerLazy()` each time defeats the purpose - the tokenizer will always be the same for this entire function (maybe it makes sense to pass it into `GetSnapshot` and store it there?)

Reply:
OK